File: platform/Android/sweeteditor/src/main/java/com/qiplat/sweeteditor/newline/NewLineActionProviderManager.java
package com.qiplat.sweeteditor.newline;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.qiplat.sweeteditor.SweetEditor;
import com.qiplat.sweeteditor.core.Document;
import com.qiplat.sweeteditor.core.foundation.TextPosition;

import java.util.ArrayList;
import java.util.List;

/**
 * Chain-based manager for newline providers, iterates until the first non-null provider takes effect.
 */
public class NewLineActionProviderManager {

    private final SweetEditor editor;
    private final List<NewLineActionProvider> providers = new ArrayList<>();

    public NewLineActionProviderManager(@NonNull SweetEditor editor) {
        this.editor = editor;
    }

    public void addProvider(@NonNull NewLineActionProvider provider) {
        providers.add(provider);
    }

    public void removeProvider(@NonNull NewLineActionProvider provider) {
        providers.remove(provider);
    }

    /**
     * Iterate all providers, return first non-null NewLineAction; return null if all return null.
     */
    @Nullable
    public NewLineAction provideNewLineAction() {
        TextPosition cursor = editor.getCursorPosition();
        if (cursor == null) return null;
        Document doc = editor.getDocument();
        String lineText = (doc != null) ? doc.getLineText(cursor.line) : "";
        if (lineText == null) lineText = "";
        NewLineContext context = new NewLineContext(
                cursor.line,
                cursor.column,
                lineText,
                editor.getLanguageConfiguration(),
                editor.getMetadata());
        for (NewLineActionProvider provider : providers) {
            NewLineAction action = provider.provideNewLineAction(context);
            if (action != null) {
                return action;
            }
        }
        return null;
    }
}

File: platform/Flutter/sweeteditor/lib/newline/newline_action_provider_manager.dart
import '../editor_types.dart';

import 'newline_types.dart';

/// Chain-based manager for newline providers.
/// Iterates until the first non-null provider takes effect.
class NewLineActionProviderManager {
  final List<NewLineActionProvider> _providers = [];

  void addProvider(NewLineActionProvider provider) {
    _providers.add(provider);
  }

  void removeProvider(NewLineActionProvider provider) {
    _providers.remove(provider);
  }

  NewLineAction? provideNewLineAction(
    int lineNumber,
    int column,
    String lineText,
    LanguageConfiguration? languageConfiguration,
    EditorMetadata? editorMetadata,
  ) {
    final context = NewLineContext(
      lineNumber: lineNumber,
      column: column,
      lineText: lineText,
      languageConfiguration: languageConfiguration,
      editorMetadata: editorMetadata,
    );
    for (final provider in _providers) {
      final action = provider.provideNewLineAction(context);
      if (action != null) return action;
    }
    return null;
  }
}
